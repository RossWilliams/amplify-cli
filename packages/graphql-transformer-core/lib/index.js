"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
require("./polyfills/Object.assign");
var TransformerContext_1 = require("./TransformerContext");
exports.TransformerContext = TransformerContext_1.default;
var Transformer_1 = require("./Transformer");
exports.Transformer = Transformer_1.default;
var GraphQLTransform_1 = require("./GraphQLTransform");
var collectDirectives_1 = require("./collectDirectives");
exports.collectDirectiveNames = collectDirectives_1.collectDirectiveNames;
exports.collectDirectivesByTypeNames = collectDirectives_1.collectDirectivesByTypeNames;
var stripDirectives_1 = require("./stripDirectives");
exports.stripDirectives = stripDirectives_1.stripDirectives;
var amplifyUtils_1 = require("./util/amplifyUtils");
exports.buildAPIProject = amplifyUtils_1.buildProject;
exports.uploadAPIProject = amplifyUtils_1.uploadDeployment;
exports.migrateAPIProject = amplifyUtils_1.migrateAPIProject;
exports.revertAPIMigration = amplifyUtils_1.revertAPIMigration;
var transformConfig_1 = require("./util/transformConfig");
exports.readProjectSchema = transformConfig_1.readSchema;
exports.readProjectConfiguration = transformConfig_1.loadProject;
exports.readTransformerConfiguration = transformConfig_1.loadConfig;
exports.writeTransformerConfiguration = transformConfig_1.writeConfig;
__export(require("./errors"));
__export(require("./util"));
exports.default = GraphQLTransform_1.default;
//# sourceMappingURL=index.js.map