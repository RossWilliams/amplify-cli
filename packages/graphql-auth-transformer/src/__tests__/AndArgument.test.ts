import { DynamoDBModelTransformer } from 'graphql-dynamodb-transformer';
import { ModelAuthTransformer } from '../ModelAuthTransformer';
import { GraphQLTransform } from 'graphql-transformer-core';

test('Test "and" parameter forces all rules to pass for read', () => {
  const validSchema = `
    type Comment @model @auth(rules: [
        {allow: groups, groups: ["Dev"], operations: [read]},
        {allow: groups, groups: ["Admin"], operations: [read], and: "testing"},
        {allow: owner, operations: [read], and: "testing"},
        {allow: groups, groupsField: "groupField", operations: [read], and: "testing"}
    ]) {
        id: ID!
        owner: ID!
        groupField: [String]!
        text: String!
        createdAt: String
        updatedAt: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [
      new DynamoDBModelTransformer(),
      new ModelAuthTransformer({
        authConfig: {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        },
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.resolvers['Query.getComment.res.vtl']).toContain('Authorization rule:');
  expect(out.resolvers['Query.listComments.res.vtl']).toContain('Authorization rule:');
  expect(out.resolvers['Query.getComment.res.vtl']).toContain('#set( $isStaticGroupAuthorized = true )');
  expect(out.resolvers['Query.getComment.res.vtl']).toContain('$util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 3');
  expect(out.resolvers['Query.listComments.res.vtl']).toContain('$util.defaultIfNull($staticCompoundAuthRuleCounts.testing, 0) == 3');
  expect(out.resolvers['Query.getComment.res.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.listComments.res.vtl']).toMatchSnapshot();
});

test('Test "create", "update", "delete" auth operations with "and" parameter forces all rules to pass', () => {
  const validSchema = `
    type Post @model @auth(rules: [
        {allow: groups, groups: ["Admin"], operations: [read, create, update, delete], and: "testing"},
        {allow: groups, groupsField: "groupField", operations: [read, create, update, delete], and: "testing"},
        {allow: owner, operations: [create, update, delete], and: "testing2"},
        {allow: groups, groups: ["Dev"], operations: [create, update, delete], and: "testing2"}
    ]) {
        id: ID!
        title: String!
        groupField: String!
        createdAt: String
        updatedAt: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [
      new DynamoDBModelTransformer(),
      new ModelAuthTransformer({
        authConfig: {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        },
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.resolvers['Query.getPost.res.vtl']).toContain('$util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 2');

  expect(out.resolvers['Query.getPost.res.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.listPosts.res.vtl']).toContain('$util.defaultIfNull($staticCompoundAuthRuleCounts.testing, 0) == 2');
  expect(out.resolvers['Query.listPosts.res.vtl']).toMatchSnapshot();

  expect(out.resolvers['Mutation.createPost.req.vtl']).toContain('$util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 2');
  expect(out.resolvers['Mutation.createPost.req.vtl']).toContain('$util.defaultIfNull($compoundAuthRuleCounts.testing2, 0) == 2');

  expect(out.resolvers['Mutation.createPost.req.vtl']).toMatchSnapshot();

  expect(out.resolvers['Mutation.updatePost.req.vtl']).toContain('#set( $innerCompoundAuth = "$innerCompoundAuth AND" )');
  expect(out.resolvers['Mutation.updatePost.req.vtl'].replace(/ +?/g, '')).toContain(
    '#if($compoundAuthRuleCounts.testing2==1)\n' + '$util.qr($compoundAuthExpressions.testing2.add("#owner0=:identity0"))\n' + '#end'
  );
  expect(out.resolvers['Mutation.updatePost.req.vtl'].replace(/ +?/g, '')).toContain(
    '#if($compoundAuthRuleCounts.testing==1)\n' +
      '$util.qr($compoundAuthExpressions.testing.add("contains(#groupsAttribute0,:group0$foreach.count)"))\n' +
      '#end'
  );

  expect(out.resolvers['Mutation.updatePost.req.vtl']).toMatchSnapshot();

  expect(out.resolvers['Mutation.deletePost.req.vtl'].replace(/ +?/g, '')).toContain(
    '#if($compoundAuthRuleCounts.testing2==1)\n' + '$util.qr($compoundAuthExpressions.testing2.add("#owner0=:identity0"))\n' + '#end'
  );
  expect(out.resolvers['Mutation.deletePost.req.vtl'].replace(/ +?/g, '')).toContain(
    '#if($compoundAuthRuleCounts.testing==1)\n' +
      '$util.qr($compoundAuthExpressions.testing.add("contains(#groupsAttribute0,:group0$foreach.count)"))\n' +
      '#end'
  );

  expect(out.resolvers['Mutation.deletePost.req.vtl']).toMatchSnapshot();
});

test('Test "update" auth operations with "and" parameter on fields prevents security bypass', () => {
  const validSchema = `
    type Post @model {
        id: ID!
        
        protectedByB: String! @auth(rules: [{allow: owner, ownerField: "id", operations: [update], and: "bRule"}])
        protectedByA: String! @auth(rules: [{allow: owner, ownerField: "id", operations: [update], and: "aRule"}])
        protectedByC: String! @auth(rules: [{allow: groups, groups: ["Admin"], operations: [update], and: "cRule"}])
        groupField: String!
        createdAt: String
        updatedAt: String
    }
    `;
  const transformer = new GraphQLTransform({
    transformers: [
      new DynamoDBModelTransformer(),
      new ModelAuthTransformer({
        authConfig: {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        },
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.resolvers['Mutation.updatePost.req.vtl']).toMatchSnapshot();
});
