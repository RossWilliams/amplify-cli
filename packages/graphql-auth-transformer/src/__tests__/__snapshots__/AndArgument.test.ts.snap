// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test "and" parameter forces all rules to pass for read 1`] = `
"#set( $compoundAuthRuleCounts = {} )
## [Start] Static Group Authorization Checks **
#set($isStaticGroupAuthorized = $util.defaultIfNull(
            $isStaticGroupAuthorized, false))
## Authorization rule: { allow: groups, groups: \\"[\\"Dev\\"]\\"   } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Dev\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $isStaticGroupAuthorized = true )
  #end
#end
## Authorization rule: { allow: groups, groups: \\"[\\"Admin\\"]\\"  , and: \\"testing\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Admin\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
  #end
#end
## [End] Static Group Authorization Checks **


## [Start] Dynamic Group Authorization Checks **
#set( $isDynamicGroupAuthorized = $util.defaultIfNull($isDynamicGroupAuthorized, false) )
## Authorization rule: { allow: groups, groupsField: \\"groupField\\"  , and: \\"testing\\" } **
#set( $allowedGroups = $util.defaultIfNull($ctx.result.groupField, []) )
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#foreach( $userGroup in $userGroups )
  #if( $util.isList($allowedGroups) )
    #if( $allowedGroups.contains($userGroup) )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
  #if( $util.isString($allowedGroups) )
    #if( $allowedGroups == $userGroup )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
#end
## [End] Dynamic Group Authorization Checks **


## [Start] Owner Authorization Checks **
#set( $isOwnerAuthorized = $util.defaultIfNull($isOwnerAuthorized, false) )
## Authorization rule: { allow: owner, ownerField: \\"owner\\", identityClaim: \\"cognito:username\\"  , and: \\"testing\\" } **
#set( $allowedOwners0 = $ctx.result.owner )
#set( $identityValue = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
#if( $util.isList($allowedOwners0) )
  #foreach( $allowedOwner in $allowedOwners0 )
    #if( $allowedOwner == $identityValue )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
#end
#if( $util.isString($allowedOwners0) )
  #if( $allowedOwners0 == $identityValue )
    #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
  #end
#end
## [End] Owner Authorization Checks **


## [Start] Throw if unauthorized **
#if( !($isStaticGroupAuthorized == true || $isDynamicGroupAuthorized == true || $isOwnerAuthorized == true || !$util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 3) )
  $util.unauthorized()
#end
## [End] Throw if unauthorized **

$util.toJson($context.result)"
`;

exports[`Test "and" parameter forces all rules to pass for read 2`] = `
"#set( $compoundAuthRuleCounts = {} )
## [Start] Static Group Authorization Checks **
#set($isStaticGroupAuthorized = $util.defaultIfNull(
            $isStaticGroupAuthorized, false))
## Authorization rule: { allow: groups, groups: \\"[\\"Dev\\"]\\"   } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Dev\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $isStaticGroupAuthorized = true )
  #end
#end
## Authorization rule: { allow: groups, groups: \\"[\\"Admin\\"]\\"  , and: \\"testing\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Admin\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
  #end
#end
## [End] Static Group Authorization Checks **


## [Start] If not static group authorized, filter items **
#if( true )
  #set( $items = [] )
  #foreach( $item in $ctx.result.items )
    ## [Start] Dynamic Group Authorization Checks **
    #set( $isLocalDynamicGroupAuthorized = false )
    ## Authorization rule: { allow: groups, groupsField: \\"groupField\\"  , and: \\"testing\\" } **
    #set( $allowedGroups = $util.defaultIfNull($item.groupField, []) )
    #set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
    #foreach( $userGroup in $userGroups )
      #if( $util.isList($allowedGroups) )
        #if( $allowedGroups.contains($userGroup) )
          #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
        #end
      #end
      #if( $util.isString($allowedGroups) )
        #if( $allowedGroups == $userGroup )
          #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
        #end
      #end
    #end
    ## [End] Dynamic Group Authorization Checks **


    ## [Start] Owner Authorization Checks **
    #set( $isLocalOwnerAuthorized = false )
    ## Authorization rule: { allow: owner, ownerField: \\"owner\\", identityClaim: \\"cognito:username\\"  , and: \\"testing\\" } **
    #set( $allowedOwners0 = $item.owner )
    #set( $identityValue = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #if( $util.isList($allowedOwners0) )
      #foreach( $allowedOwner in $allowedOwners0 )
        #if( $allowedOwner == $identityValue )
          #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
        #end
      #end
    #end
    #if( $util.isString($allowedOwners0) )
      #if( $allowedOwners0 == $identityValue )
        #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
      #end
    #end
    ## [End] Owner Authorization Checks **


    #if( $isStaticGroupAuthorized == true || $isLocalDynamicGroupAuthorized == true || $isLocalOwnerAuthorized == true || $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 3 )
      $util.qr($items.add($item))
    #end
  #end
  #set( $ctx.result.items = $items )
#end
## [End] If not static group authorized, filter items **

$util.toJson($ctx.result)"
`;

exports[`Test "create", "update", "delete" auth operations with "and" parameter forces all rules to pass 1`] = `
"#set( $compoundAuthRuleCounts = {} )
## [Start] Static Group Authorization Checks **
#set($isStaticGroupAuthorized = $util.defaultIfNull(
            $isStaticGroupAuthorized, false))
## Authorization rule: { allow: groups, groups: \\"[\\"Admin\\"]\\"  , and: \\"testing\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Admin\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
  #end
#end
## [End] Static Group Authorization Checks **


## [Start] Dynamic Group Authorization Checks **
#set( $isDynamicGroupAuthorized = $util.defaultIfNull($isDynamicGroupAuthorized, false) )
## Authorization rule: { allow: groups, groupsField: \\"groupField\\"  , and: \\"testing\\" } **
#set( $allowedGroups = $util.defaultIfNull($ctx.result.groupField, []) )
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#foreach( $userGroup in $userGroups )
  #if( $util.isList($allowedGroups) )
    #if( $allowedGroups.contains($userGroup) )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
  #if( $util.isString($allowedGroups) )
    #if( $allowedGroups == $userGroup )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
#end
## [End] Dynamic Group Authorization Checks **


## No Owner Authorization Rules **


## [Start] Throw if unauthorized **
#if( !($isStaticGroupAuthorized == true || $isDynamicGroupAuthorized == true || $isOwnerAuthorized == true || !$util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 2) )
  $util.unauthorized()
#end
## [End] Throw if unauthorized **

$util.toJson($context.result)"
`;

exports[`Test "create", "update", "delete" auth operations with "and" parameter forces all rules to pass 2`] = `
"#set( $compoundAuthRuleCounts = {} )
## [Start] Static Group Authorization Checks **
#set($isStaticGroupAuthorized = $util.defaultIfNull(
            $isStaticGroupAuthorized, false))
## Authorization rule: { allow: groups, groups: \\"[\\"Admin\\"]\\"  , and: \\"testing\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Admin\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
  #end
#end
## [End] Static Group Authorization Checks **


## [Start] If not static group authorized, filter items **
#if( true )
  #set( $items = [] )
  #foreach( $item in $ctx.result.items )
    ## [Start] Dynamic Group Authorization Checks **
    #set( $isLocalDynamicGroupAuthorized = false )
    ## Authorization rule: { allow: groups, groupsField: \\"groupField\\"  , and: \\"testing\\" } **
    #set( $allowedGroups = $util.defaultIfNull($item.groupField, []) )
    #set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
    #foreach( $userGroup in $userGroups )
      #if( $util.isList($allowedGroups) )
        #if( $allowedGroups.contains($userGroup) )
          #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
        #end
      #end
      #if( $util.isString($allowedGroups) )
        #if( $allowedGroups == $userGroup )
          #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
        #end
      #end
    #end
    ## [End] Dynamic Group Authorization Checks **


    ## No Owner Authorization Rules **


    #if( $isStaticGroupAuthorized == true || $isLocalDynamicGroupAuthorized == true || $isLocalOwnerAuthorized == true || $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 2 )
      $util.qr($items.add($item))
    #end
  #end
  #set( $ctx.result.items = $items )
#end
## [End] If not static group authorized, filter items **

$util.toJson($ctx.result)"
`;

exports[`Test "create", "update", "delete" auth operations with "and" parameter forces all rules to pass 3`] = `
"#set( $compoundAuthRuleCounts = {} )
## [Start] Static Group Authorization Checks **
#set($isStaticGroupAuthorized = $util.defaultIfNull(
            $isStaticGroupAuthorized, false))
## Authorization rule: { allow: groups, groups: \\"[\\"Admin\\"]\\"  , and: \\"testing\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Admin\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
  #end
#end
## Authorization rule: { allow: groups, groups: \\"[\\"Dev\\"]\\"  , and: \\"testing2\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $allowedGroups = [\\"Dev\\"] )
#foreach( $userGroup in $userGroups )
  #if( $allowedGroups.contains($userGroup) )
    #set( $compoundAuthRuleCounts.testing2 = $util.defaultIfNull($compoundAuthRuleCounts.testing2, 0) + 1 )
  #end
#end
## [End] Static Group Authorization Checks **


## [Start] Dynamic Group Authorization Checks **
## Authorization rule: { allow: groups, groupsField: \\"groupField\\"  , and: \\"testing\\" } **
#set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:groups\\"), []) )
#set( $isDynamicGroupAuthorized = $util.defaultIfNull($isDynamicGroupAuthorized, false) )
#foreach( $userGroup in $userGroups )
  #if( $util.isList($ctx.args.input.groupField) )
    #if( $ctx.args.input.groupField.contains($userGroup) )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
  #if( $util.isString($ctx.args.input.groupField) )
    #if( $ctx.args.input.groupField == $userGroup )
      #set( $compoundAuthRuleCounts.testing = $util.defaultIfNull($compoundAuthRuleCounts.testing, 0) + 1 )
    #end
  #end
#end
## [End] Dynamic Group Authorization Checks **


## [Start] Owner Authorization Checks **
#set( $isOwnerAuthorized = false )
## Authorization rule: { allow: owner, ownerField: \\"owner\\", identityClaim: \\"cognito:username\\"  , and: \\"testing2\\" } **
#set( $allowedOwners0 = $util.defaultIfNull($ctx.args.input.owner, null) )
#set( $identityValue = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
#if( $util.isList($allowedOwners0) )
  #foreach( $allowedOwner in $allowedOwners0 )
    #if( $allowedOwner == $identityValue )
      #set( $compoundAuthRuleCounts.testing2 = $util.defaultIfNull($compoundAuthRuleCounts.testing2, 0) + 1 )
    #end
  #end
#end
#if( $util.isString($allowedOwners0) )
  #if( $allowedOwners0 == $identityValue )
    #set( $compoundAuthRuleCounts.testing2 = $util.defaultIfNull($compoundAuthRuleCounts.testing2, 0) + 1 )
  #end
#end
#if( $util.isNull($allowedOwners0) && (! $ctx.args.input.containsKey(\\"owner\\")) )
  $util.qr($ctx.args.input.put(\\"owner\\", $identityValue))
  #set( $compoundAuthRuleCounts.testing2 = $util.defaultIfNull($compoundAuthRuleCounts.testing2, 0) + 1 )
#end
## [End] Owner Authorization Checks **


## [Start] Throw if unauthorized **
#if( !($isStaticGroupAuthorized == true || $isDynamicGroupAuthorized == true || $isOwnerAuthorized == true || !$util.defaultIfNull($compoundAuthRuleCounts.testing, 0) == 2 || !$util.defaultIfNull($compoundAuthRuleCounts.testing2, 0) == 2) )
  $util.unauthorized()
#end
## [End] Throw if unauthorized **

## [Start] Prepare DynamoDB PutItem Request. **
$util.qr($context.args.input.put(\\"createdAt\\", $util.defaultIfNull($ctx.args.input.createdAt, $util.time.nowISO8601())))
$util.qr($context.args.input.put(\\"updatedAt\\", $util.defaultIfNull($ctx.args.input.updatedAt, $util.time.nowISO8601())))
$util.qr($context.args.input.put(\\"__typename\\", \\"Post\\"))
{
  \\"version\\": \\"2017-02-28\\",
  \\"operation\\": \\"PutItem\\",
  \\"key\\": #if( $modelObjectKey ) $util.toJson($modelObjectKey) #else {
  \\"id\\":   $util.dynamodb.toDynamoDBJson($util.defaultIfNullOrBlank($ctx.args.input.id, $util.autoId()))
} #end,
  \\"attributeValues\\": $util.dynamodb.toMapValuesJson($context.args.input),
  \\"condition\\": {
      \\"expression\\": \\"attribute_not_exists(#id)\\",
      \\"expressionNames\\": {
          \\"#id\\": \\"id\\"
    }
  }
}
## [End] Prepare DynamoDB PutItem Request. **"
`;
